File: backend/app/services/documents.py
"""Document upload, listing, deletion, and durable ingestion services."""

from __future__ import annotations

import asyncio
from typing import Any
from uuid import UUID, uuid4

from sqlalchemy.ext.asyncio import AsyncSession, async_sessionmaker

from app.core.config import Settings
from app.core.logging import (
    get_logger,
    log_process_failed,
    log_process_finished,
    log_process_started,
    request_id_context,
)
from app.db.tenancy import set_tenant_context
from app.exceptions import NotFoundError, ValidationError
from app.integrations.embeddings import EmbeddingService
from app.integrations.qdrant import QdrantVectorStore
from app.integrations.supabase_storage import SupabaseStorage
from app.models.entities import Document
from app.models.enums import DocumentStatus
from app.repositories.documents import DocumentRepository
from app.repositories.usage import AuditRepository, UsageRepository
from app.services.chunking import DocumentChunker
from app.services.extraction import extract_document
from app.utils.files import ValidatedUpload

logger = get_logger(__name__)


class DocumentService:
    """Request-facing document operations."""

    def __init__(
        self,
        session: AsyncSession,
        settings: Settings,
        storage: SupabaseStorage,
        vector_store: QdrantVectorStore,
    ) -> None:
        self.session = session
        self.settings = settings
        self.storage = storage
        self.vector_store = vector_store
        self.documents = DocumentRepository(session)
        self.usage = UsageRepository(session)
        self.audit = AuditRepository(session)

    async def upload(self, user_id: UUID, upload: ValidatedUpload) -> Document:
        """Persist upload state, store bytes, and enqueue via PostgreSQL status."""
        log_process_started(
            logger,
            "Upload document",
            user_id=str(user_id),
            filename=upload.safe_filename,
            bytes=len(upload.content),
        )
        subscription = await self.usage.get_subscription(user_id, for_update=True)
        usage_limits = subscription.usage_limits if subscription else {}
        storage_limit = int(usage_limits.get("storage_bytes", 100 * 1024 * 1024))
        current_storage = await self.documents.total_storage_bytes(user_id)
        if current_storage + len(upload.content) > storage_limit:
            raise ValidationError("Your plan's storage limit has been reached.")
        document_id = uuid4()
        storage_path = f"{user_id}/{document_id}/{uuid4().hex}-{upload.safe_filename}"
        log_process_started(
            logger,
            "Create document database record",
            user_id=str(user_id),
            document_id=str(document_id),
        )
        document = await self.documents.create(
            id=document_id,
            user_id=user_id,
            filename=upload.safe_filename,
            original_filename=upload.original_filename,
            content_type=upload.content_type,
            file_size=len(upload.content),
            checksum_sha256=upload.checksum_sha256,
            storage_bucket=self.settings.supabase_storage_bucket,
            storage_path=storage_path,
            status=DocumentStatus.UPLOADING,
        )
        await self.session.commit()
        log_process_finished(
            logger,
            "Create document database record",
            user_id=str(user_id),
            document_id=str(document_id),
        )
        try:
            await self.storage.upload(
                bucket=document.storage_bucket,
                path=document.storage_path,
                content=upload.content,
                content_type=document.content_type,
            )
            await set_tenant_context(self.session, user_id)
            document = await self.documents.get(user_id, document_id, for_update=True)
            if document is None:
                raise NotFoundError("Document record disappeared during upload.")
            log_process_started(
                logger,
                "Mark document processing",
                user_id=str(user_id),
                document_id=str(document_id),
            )
            await self.documents.mark_processing(document)
            await self.usage.increment(
                user_id, documents_uploaded=1, storage_bytes=document.file_size
            )
            await self.audit.record(
                user_id=user_id,
                action="document.uploaded",
                resource_type="document",
                resource_id=document.id,
                request_id=request_id_context.get(),
                event_metadata={
                    "filename": document.filename,
                    "bytes": document.file_size,
                },
            )
            # On-update SQL expressions expire their ORM attributes after flush. Reload them
            # while the async session is active so response serialization cannot trigger I/O.
            await self.session.refresh(document)
            await self.session.commit()
            log_process_finished(
                logger,
                "Mark document processing",
                user_id=str(user_id),
                document_id=str(document_id),
            )
        except Exception as exc:
            log_process_failed(
                logger,
                "Upload document",
                user_id=str(user_id),
                document_id=str(document_id),
            )
            await self.session.rollback()
            await set_tenant_context(self.session, user_id)
            failed = await self.documents.get(user_id, document_id, for_update=True)
            if failed is not None:
                await self.documents.mark_failed(failed, "Storage upload failed.")
                await self.session.commit()
            raise exc
        logger.info(
            "document_upload_accepted — file stored in Supabase, marked PROCESSING, worker will pick it up",
            document_id=str(document.id),
            user_id=str(user_id),
            filename=document.filename,
            bytes=document.file_size,
        )
        log_process_finished(
            logger,
            "Upload document",
            user_id=str(user_id),
            document_id=str(document.id),
            filename=document.filename,
        )
        return document

    async def delete(self, user_id: UUID, document_id: UUID) -> None:
        """Delete external data and then soft-delete tenant metadata."""
        log_process_started(
            logger,
            "Delete document",
            user_id=str(user_id),
            document_id=str(document_id),
        )
        document = await self.documents.get(user_id, document_id, for_update=True)
        if document is None:
            raise NotFoundError("Document not found.")
        await self.storage.delete(bucket=document.storage_bucket, path=document.storage_path)
        await self.vector_store.delete_document(user_id=user_id, document_id=document.id)
        await self.documents.soft_delete(document)
        await self.audit.record(
            user_id=user_id,
            action="document.deleted",
            resource_type="document",
            resource_id=document.id,
            request_id=request_id_context.get(),
        )
        await self.session.commit()
        log_process_finished(
            logger,
            "Delete document",
            user_id=str(user_id),
            document_id=str(document_id),
        )
        logger.info(
            "document_deleted — removed from storage, vectors purged, record soft-deleted",
            document_id=str(document_id),
            user_id=str(user_id),
        )


class DocumentProcessor:
    """Queue-free durable processor claimed through PostgreSQL row locks."""

    def __init__(
        self,
        *,
        session_factory: async_sessionmaker[AsyncSession],
        settings: Settings,
        storage: SupabaseStorage,
        embeddings: EmbeddingService,
        vector_store: QdrantVectorStore,
    ) -> None:
        self.session_factory = session_factory
        self.settings = settings
        self.storage = storage
        self.embeddings = embeddings
        self.vector_store = vector_store
        self.chunker = DocumentChunker(settings.chunk_size, settings.chunk_overlap)

    async def process_one(self, document_id: UUID | None = None) -> bool:
        """Claim and process one document; return whether work was found."""
        async with self.session_factory() as session:
            repository = DocumentRepository(session)
            document = await repository.claim(
                document_id=document_id,
                stale_minutes=self.settings.worker_stale_minutes,
                max_retries=self.settings.worker_max_retries,
            )
            if document is None:
                await session.rollback()
                return False
            claimed = {
                "id": document.id,
                "user_id": document.user_id,
                "bucket": document.storage_bucket,
                "path": document.storage_path,
                "filename": document.filename,
                "created_at": document.created_at,
                "retry_count": document.retry_count,
            }
            await session.commit()

        logger.info(
            "document_claimed — starting extraction pipeline",
            document_id=str(claimed["id"]),
            user_id=str(claimed["user_id"]),
            filename=claimed["filename"],
            retry_count=claimed["retry_count"],
        )
        log_process_started(
            logger,
            "Process document",
            document_id=str(claimed["id"]),
            user_id=str(claimed["user_id"]),
            filename=claimed["filename"],
            retry_count=claimed["retry_count"],
        )
        try:
            content = await self.storage.download(
                bucket=str(claimed["bucket"]), path=str(claimed["path"])
            )
            logger.info(
                "document_download_done — bytes fetched from Supabase Storage, running extractor",
                document_id=str(claimed["id"]),
                bytes=len(content),
                filename=claimed["filename"],
            )
            log_process_started(
                logger,
                "Extract text",
                document_id=str(claimed["id"]),
                filename=claimed["filename"],
            )
            log_process_started(
                logger,
                "Extract metadata",
                document_id=str(claimed["id"]),
                filename=claimed["filename"],
            )
            extraction = await asyncio.to_thread(
                extract_document, str(claimed["filename"]), content
            )
            log_process_finished(
                logger,
                "Extract text",
                document_id=str(claimed["id"]),
                pages=extraction.page_count,
                sections=len(extraction.sections),
            )
            log_process_finished(
                logger,
                "Extract metadata",
                document_id=str(claimed["id"]),
                metadata_fields=len(extraction.metadata),
            )
            logger.info(
                "document_extracted — text pulled, splitting into chunks",
                document_id=str(claimed["id"]),
                pages=extraction.page_count,
            )
            log_process_started(
                logger,
                "Chunk text",
                document_id=str(claimed["id"]),
                sections=len(extraction.sections),
            )
            text_chunks = self.chunker.split(extraction.sections)
            if not text_chunks:
                raise ValidationError("No chunks were produced from this document.")
            log_process_finished(
                logger,
                "Chunk text",
                document_id=str(claimed["id"]),
                chunks=len(text_chunks),
            )
            logger.info(
                "document_chunked — generating embeddings via Ollama",
                document_id=str(claimed["id"]),
                chunk_count=len(text_chunks),
            )
            vectors = await self.embeddings.embed_documents(
                [chunk.chunk_text for chunk in text_chunks]
            )
            logger.info(
                "embeddings_ready — upserting chunks to Postgres and vectors to Qdrant",
                document_id=str(claimed["id"]),
                vector_count=len(vectors),
            )
            rows: list[Any]
            log_process_started(
                logger,
                "Store document chunks",
                document_id=str(claimed["id"]),
                chunks=len(text_chunks),
            )
            async with self.session_factory() as session:
                await set_tenant_context(session, claimed["user_id"])
                repository = DocumentRepository(session)
                document = await repository.get(claimed["user_id"], claimed["id"], for_update=True)
                if document is None or document.deleted_at is not None:
                    await session.rollback()
                    return True
                rows = await repository.replace_chunks(
                    document,
                    [
                        {
                            "chunk_index": chunk.chunk_index,
                            "page_number": chunk.page_number,
                            "chunk_text": chunk.chunk_text,
                            "token_count": chunk.token_count,
                            "qdrant_point_id": uuid4(),
                            "chunk_metadata": chunk.chunk_metadata,
                        }
                        for chunk in text_chunks
                    ],
                )
                await session.commit()
            log_process_finished(
                logger,
                "Store document chunks",
                document_id=str(claimed["id"]),
                chunks=len(rows),
            )
            await self.vector_store.ensure_collection()
            await self.vector_store.delete_document(
                user_id=claimed["user_id"], document_id=claimed["id"]
            )
            await self.vector_store.upsert_chunks(
                chunks=rows,
                vectors=vectors,
                filename=str(claimed["filename"]),
                created_at=claimed["created_at"],
            )
            log_process_started(
                logger,
                "Mark document ready",
                document_id=str(claimed["id"]),
                chunks=len(rows),
            )
            async with self.session_factory() as session:
                await set_tenant_context(session, claimed["user_id"])
                repository = DocumentRepository(session)
                document = await repository.get(claimed["user_id"], claimed["id"], for_update=True)
                if document is not None:
                    await repository.mark_ready(
                        document,
                        page_count=extraction.page_count,
                        chunk_count=len(rows),
                        metadata=extraction.metadata,
                    )
                    await session.commit()
            log_process_finished(
                logger,
                "Mark document ready",
                document_id=str(claimed["id"]),
                pages=extraction.page_count,
                chunks=len(rows),
            )
            logger.info(
                "document_processing_completed — document is READY and fully searchable",
                document_id=str(claimed["id"]),
                user_id=str(claimed["user_id"]),
                chunks=len(rows),
                pages=extraction.page_count,
            )
            log_process_finished(
                logger,
                "Process document",
                document_id=str(claimed["id"]),
                user_id=str(claimed["user_id"]),
                pages=extraction.page_count,
                chunks=len(rows),
            )
        except Exception as exc:
            log_process_failed(
                logger,
                "Process document",
                document_id=str(claimed["id"]),
                user_id=str(claimed["user_id"]),
                attempt=claimed["retry_count"],
            )
            logger.exception(
                "document_processing_failed — will retry or mark FAILED if max retries exceeded",
                document_id=str(claimed["id"]),
                user_id=str(claimed["user_id"]),
                attempt=claimed["retry_count"],
                error=str(exc),
            )
            async with self.session_factory() as session:
                await set_tenant_context(session, claimed["user_id"])
                repository = DocumentRepository(session)
                document = await repository.get(claimed["user_id"], claimed["id"], for_update=True)
                if document is not None:
                    safe_error = f"{type(exc).__name__}: {str(exc)}"
                    if document.retry_count >= self.settings.worker_max_retries:
                        await repository.mark_failed(document, safe_error)
                    else:
                        await repository.release_for_retry(document, safe_error)
                    await session.commit()
        return True

File: backend/src/services/documents.js
import crypto from "node:crypto";
import { config } from "../config.js";
import { withTransaction } from "../db.js";
import { NotFoundError, ValidationError } from "../errors.js";
import { embeddings } from "../integrations/ollama.js";
import { vectorStore } from "../integrations/qdrant.js";
import { storage } from "../integrations/storage.js";
import { log, logProcessFailed, logProcessFinished, logProcessStarted } from "../logger.js";
import { getSubscription, incrementUsage, recordAudit } from "../repositories/account.js";
import {
  claimDocument,
  createDocument,
  getDocument,
  markDocumentFailed,
  markDocumentProcessing,
  markDocumentReady,
  publicDocument,
  releaseDocumentForRetry,
  replaceDocumentChunks,
  softDeleteDocument,
  totalStorageBytes,
} from "../repositories/documents.js";
import { DocumentChunker } from "./chunking.js";
import { extractDocument } from "./extraction.js";

export class DocumentService {
  async upload({ userId, upload, requestId }) {
    logProcessStarted("Upload document", {
      user_id: userId,
      original_filename: upload.originalFilename,
      safe_filename: upload.safeFilename,
      bytes: upload.content.length,
    });
    const documentId = crypto.randomUUID();
    const storagePath = `${userId}/${documentId}/${crypto.randomBytes(16).toString("hex")}-${upload.safeFilename}`;

    const document = await withTransaction(
      async (client) => {
        const subscription = await getSubscription(client, userId, { forUpdate: true });
        const storageLimit = Number(subscription?.usage_limits?.storage_bytes ?? 100 * 1024 * 1024);
        const currentStorage = await totalStorageBytes(client, userId);
        if (currentStorage + upload.content.length > storageLimit) {
          throw new ValidationError("Your plan's storage limit has been reached.");
        }
        log("info", "document_upload_storage_limit_checked", {
          message: "Document upload storage limit checked.",
          user_id: userId,
          current_storage_bytes: currentStorage,
          upload_bytes: upload.content.length,
          storage_limit_bytes: storageLimit,
        });
        return createDocument(client, {
          id: documentId,
          userId,
          filename: upload.safeFilename,
          originalFilename: upload.originalFilename,
          contentType: upload.contentType,
          fileSize: upload.content.length,
          checksumSha256: upload.checksumSha256,
          storageBucket: config.supabaseStorageBucket,
          storagePath,
          status: "UPLOADING",
        });
      },
      { userId },
    );

    try {
      log("info", "document_storage_upload_started", {
        message: "Document bytes upload to private storage started.",
        document_id: documentId,
        user_id: userId,
        bucket: document.storage_bucket,
        path: document.storage_path,
        bytes: upload.content.length,
      });
      await storage.upload({
        bucket: document.storage_bucket,
        path: document.storage_path,
        content: upload.content,
        contentType: document.content_type,
      });
      log("info", "document_storage_upload_finished", {
        message: "Document bytes uploaded to private storage.",
        document_id: documentId,
        user_id: userId,
      });
      const processing = await withTransaction(
        async (client) => {
          const current = await getDocument(client, userId, documentId, { forUpdate: true });
          if (!current) {
            throw new NotFoundError("Document record disappeared during upload.");
          }
          const updated = await markDocumentProcessing(client, documentId);
          await incrementUsage(client, userId, {
            documentsUploaded: 1,
            storageBytes: Number(document.file_size),
          });
          await recordAudit(client, {
            userId,
            action: "document.uploaded",
            resourceType: "document",
            resourceId: documentId,
            requestId,
            metadata: { filename: document.filename, bytes: Number(document.file_size) },
          });
          return updated;
        },
        { userId },
      );
      logProcessFinished("Upload document", {
        document_id: documentId,
        user_id: userId,
        status: processing.status,
      });
      return publicDocument(processing);
    } catch (error) {
      logProcessFailed("Upload document", error, {
        document_id: documentId,
        user_id: userId,
      });
      await withTransaction(
        async (client) => {
          const current = await getDocument(client, userId, documentId, { forUpdate: true });
          if (current) {
            await markDocumentFailed(client, documentId, "Storage upload failed.");
          }
        },
        { userId },
      ).catch(() => undefined);
      throw error;
    }
  }

  async delete({ userId, documentId, requestId }) {
    logProcessStarted("Delete document", { user_id: userId, document_id: documentId });
    const document = await withTransaction(
      async (client) => {
        const result = await getDocument(client, userId, documentId, { forUpdate: true });
        if (!result) {
          throw new NotFoundError("Document not found.");
        }
        return result;
      },
      { userId },
    );
    await storage.delete({ bucket: document.storage_bucket, path: document.storage_path });
    await vectorStore.deleteDocument({ userId, documentId });
    await withTransaction(
      async (client) => {
        await softDeleteDocument(client, documentId);
        await recordAudit(client, {
          userId,
          action: "document.deleted",
          resourceType: "document",
          resourceId: documentId,
          requestId,
        });
      },
      { userId },
    );
    logProcessFinished("Delete document", { user_id: userId, document_id: documentId });
  }
}

export class DocumentProcessor {
  constructor() {
    this.chunker = new DocumentChunker(config.chunkSize, config.chunkOverlap);
  }

  async processOne(documentId = null) {
    const claimed = await withTransaction((client) =>
      claimDocument(client, {
        documentId,
        staleMinutes: config.workerStaleMinutes,
        maxRetries: config.workerMaxRetries,
      }),
    );
    if (!claimed) {
      return false;
    }

    logProcessStarted("Process document", {
      document_id: claimed.id,
      user_id: claimed.user_id,
      retry_count: claimed.retry_count,
    });
    try {
      log("info", "document_processing_download_started", {
        message: "Document processing download started.",
        document_id: claimed.id,
        user_id: claimed.user_id,
        bucket: claimed.storage_bucket,
        path: claimed.storage_path,
      });
      const content = await storage.download({
        bucket: claimed.storage_bucket,
        path: claimed.storage_path,
      });
      log("info", "document_processing_download_finished", {
        message: "Document processing download finished.",
        document_id: claimed.id,
        user_id: claimed.user_id,
        bytes: content.length,
      });
      const extraction = await extractDocument(claimed.filename, content);
      log("info", "document_text_extracted", {
        message: "Document text extracted.",
        document_id: claimed.id,
        user_id: claimed.user_id,
        sections: extraction.sections.length,
        page_count: extraction.pageCount,
        metadata: extraction.metadata,
      });
      const textChunks = this.chunker.split(extraction.sections);
      log("info", "document_chunks_created", {
        message: "Document chunks created.",
        document_id: claimed.id,
        user_id: claimed.user_id,
        chunks: textChunks.length,
        chunk_size: config.chunkSize,
        chunk_overlap: config.chunkOverlap,
      });
      if (textChunks.length === 0) {
        throw new ValidationError("No chunks were produced from this document.");
      }
      const vectors = await embeddings.embedDocuments(textChunks.map((chunk) => chunk.chunkText));
      log("info", "document_embeddings_created", {
        message: "Document embeddings created.",
        document_id: claimed.id,
        user_id: claimed.user_id,
        vectors: vectors.length,
        dimensions: vectors[0]?.length ?? 0,
      });
      const rows = await withTransaction(
        async (client) => {
          const current = await getDocument(client, claimed.user_id, claimed.id, {
            forUpdate: true,
          });
          if (!current) {
            return [];
          }
          return replaceDocumentChunks(client, current, textChunks);
        },
        { userId: claimed.user_id },
      );
      log("info", "document_chunks_saved", {
        message: "Document chunks saved to PostgreSQL.",
        document_id: claimed.id,
        user_id: claimed.user_id,
        chunks: rows.length,
      });
      if (rows.length === 0) {
        return true;
      }
      await vectorStore.ensureCollection();
      await vectorStore.deleteDocument({ userId: claimed.user_id, documentId: claimed.id });
      await vectorStore.upsertChunks({
        chunks: rows,
        vectors,
        filename: claimed.filename,
        createdAt: claimed.created_at,
      });
      log("info", "document_vectors_saved", {
        message: "Document vectors saved to Qdrant.",
        document_id: claimed.id,
        user_id: claimed.user_id,
        vectors: rows.length,
      });
      await withTransaction(
        (client) =>
          markDocumentReady(client, claimed.id, {
            pageCount: extraction.pageCount,
            chunkCount: rows.length,
            metadata: extraction.metadata,
          }),
        { userId: claimed.user_id },
      );
      logProcessFinished("Process document", {
        document_id: claimed.id,
        user_id: claimed.user_id,
        chunks: rows.length,
      });
    } catch (error) {
      logProcessFailed("Process document", error, {
        document_id: claimed.id,
        user_id: claimed.user_id,
      });
      await withTransaction(
        async (client) => {
          const current = await getDocument(client, claimed.user_id, claimed.id, {
            forUpdate: true,
          });
          if (!current) {
            return;
          }
          const safeError = `${error.constructor?.name ?? "Error"}: ${error.message}`;
          if (current.retry_count >= config.workerMaxRetries) {
            await markDocumentFailed(client, claimed.id, safeError);
          } else {
            await releaseDocumentForRetry(client, claimed.id, safeError);
          }
        },
        { userId: claimed.user_id },
      );
    }
    return true;
  }
}

export const documentService = new DocumentService();
export const documentProcessor = new DocumentProcessor();
